### PR TITLE
Adding FA2 support for MusicGen

### LINKED ISSUE #27552: Flash Attention 2 for audio/musicgen
### Feature request

Add flash attention 2 to musicgen models (and or audio gen)

### Motivation

Musicgen is a *really* large model, and its very hard to run on consumer GPUs, adding flash attention could make the model much lighter and possibly faster.

### Your contribution

I can test on both WSL and windows 11 on a 3060

### PR DESCRIPTION
**Pull request by staghado**
# What does this PR do?

This PR adds Flash Attention 2 support for MusicGen model. It is based on Bart example and it is a WIP for now.
I could not test the model because FA2 is not supported yet for T4 GPUs.

Fixes #27552

@sanchit-gandhi @ylacombe

### CODE REVIEW

**Comment by ylacombe** [comment]
Hey @staghado, thanks for taking care of this, let us know when it's ready to be reviewed!

**Review by circuluspibo — APPROVED** [review]
working test

**Comment by staghado** [comment]
I have conducted some tests on an A10 GPU : 
    - The code seems to work without errors when `_supports_flash_attn_2` is set to `True` for `MusicgenForConditionalGeneration` but does not load the model with FA2 if not specified by hand. Maybe it needs to be added at the class level in MusicgenForConditionalGeneration?
    - There is no difference in generation speed between eager attention and FA2 : 
    
Screenshot from 2024-01-06 21-50-49

**Comment by sanchit-gandhi** [comment]
cc @ylacombe could you possibly circle back here when you get the chance!

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 1]
```diff
@@ -22,13 +22,20 @@
...
+import torch.nn.functional as F
 from torch.nn import CrossEntropyLoss
 
 from ...activations import ACT2FN
 from ...generation.configuration_utils import GenerationConfig
-from ...generation.logits_process import ClassifierFreeGuidanceLogitsProcessor, LogitsProcessorList
+from ...generation.logits_process import (
+    ClassifierFreeGuidanceLogitsProcessor,
+    LogitsProcessorList,
+)
```
Comment (ylacombe) [inline src/transformers/models/musicgen/modeling_musicgen.py:33]: Please remove the final comma and put in the same lines !
```suggestion
from ...generation.logits_process import ClassifierFreeGuidanceLogitsProcessor, LogitsProcessorList
```

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 2]
```diff
@@ -22,13 +22,20 @@
...
 from ...generation.stopping_criteria import StoppingCriteriaList
-from ...modeling_attn_mask_utils import _prepare_4d_attention_mask, _prepare_4d_causal_attention_mask
+from ...modeling_attn_mask_utils import (
+    _prepare_4d_attention_mask,
+    _prepare_4d_causal_attention_mask,
+)
```
Comment (ylacombe) [inline src/transformers/models/musicgen/modeling_musicgen.py:38]: Same here
```suggestion
from ...modeling_attn_mask_utils import _prepare_4d_attention_mask, _prepare_4d_causal_attention_mask
```

**Review by ylacombe — COMMENTED** [review]
Hi @staghado, thanks for the update here!

I believe that you have to add `_supports_flash_attn_2` to `MusicgenForConditionalGeneration` and `MusicgenForCausalLM`, otherwise it won't be flagged as supporting FA2 !

Also, did you make sure to add the flag `attn_implementation="flash_attention_2"` to `from_pretrained` ? as indicated [here](https://huggingface.co/docs/transformers/v4.36.1/en/perf_infer_gpu_one#flashattention-2). Let me know.


Besides that, the modeling code looks good except some changes of format that shouldn't happen where you add a coma at the end of the line, all along the modeling file.

You can also add Muscigen to the list of models supported [here](https://huggingface.co/docs/transformers/v4.36.1/en/perf_infer_gpu_one#flashattention-2).

Let me know if you need further help !

**Comment by staghado** [comment]
Hi @ylacombe,

I confirm that the model was instantiated as described here : [here](https://huggingface.co/docs/transformers/v4.36.1/en/perf_infer_gpu_one#flashattention-2) with the exception of `torch_dtype=torch.float16` instead of `torch_dtype=torch.bfloat16` because some operations did not seem to implement bfloat16.

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 3]
```diff
@@ -818,7 +1071,13 @@ def forward(
...
             return tuple(
                 v
-                for v in [hidden_states, next_cache, all_hidden_states, all_self_attns, all_cross_attentions]
+                for v in [
+                    hidden_states,
+                    next_cache,
+                    all_hidden_states,
+                    all_self_attns,
+                    all_cross_attentions,
+                ]
```
Comment (ylacombe) [inline src/transformers/models/musicgen/modeling_musicgen.py:1080]: Should be a one liner
```suggestion
                for v in [hidden_states, next_cache, all_hidden_states, all_self_attns, all_cross_attentions]
```

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 4]
```diff
@@ -1074,7 +1333,12 @@ def build_delay_pattern_mask(self, input_ids: torch.LongTensor, pad_token_id: in
 
         max_length = max_length if max_length is not None else self.generation_config.max_length
         input_ids_shifted = (
-            torch.ones((bsz, num_codebooks, max_length), dtype=torch.long, device=input_ids.device) * -1
+            torch.ones(
+                (bsz, num_codebooks, max_length),
+                dtype=torch.long,
+                device=input_ids.device,
+            )
+            * -1
```
Comment (ylacombe) [inline src/transformers/models/musicgen/modeling_musicgen.py:1341]: Same

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 5]
```diff
@@ -1095,7 +1359,8 @@ def build_delay_pattern_mask(self, input_ids: torch.LongTensor, pad_token_id: in
         # construct a pattern mask that indicates the positions of padding tokens for each codebook
         # first fill the upper triangular part (the EOS padding)
         delay_pattern = torch.triu(
-            torch.ones((channel_codebooks, max_length), dtype=torch.bool), diagonal=max_length - channel_codebooks + 1
+            torch.ones((channel_codebooks, max_length), dtype=torch.bool),
+            diagonal=max_length - channel_codebooks + 1,
```
Comment (ylacombe) [inline src/transformers/models/musicgen/modeling_musicgen.py:1363]: Same

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 6]
```diff
@@ -1252,7 +1518,9 @@ def generate(
         requires_attention_mask = "encoder_outputs" not in model_kwargs
         if model_kwargs.get("attention_mask", None) is None and requires_attention_mask:
             model_kwargs["attention_mask"] = self._prepare_attention_mask_for_generation(
-                input_ids, generation_config.pad_token_id, generation_config.eos_token_id
+                input_ids,
+                generation_config.pad_token_id,
+                generation_config.eos_token_id,
```
Comment (ylacombe) [inline src/transformers/models/musicgen/modeling_musicgen.py:1523]: Same

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 7]
```diff
@@ -1510,7 +1779,9 @@ def tie_weights(self):
             # tie text encoder and decoder base model
             decoder_base_model_prefix = self.decoder.base_model_prefix
             self._tie_encoder_decoder_weights(
-                self.text_encoder, self.decoder._modules[decoder_base_model_prefix], self.decoder.base_model_prefix
+                self.text_encoder,
+                self.decoder._modules[decoder_base_model_prefix],
+                self.decoder.base_model_prefix,
```
Comment (ylacombe) [inline src/transformers/models/musicgen/modeling_musicgen.py:1784]: Same

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 8]
```diff
@@ -1671,7 +1942,9 @@ def from_sub_models_pretrained(
 
             if "config" not in kwargs_text_encoder:
                 encoder_config, kwargs_text_encoder = AutoConfig.from_pretrained(
-                    text_encoder_pretrained_model_name_or_path, **kwargs_text_encoder, return_unused_kwargs=True
+                    text_encoder_pretrained_model_name_or_path,
+                    **kwargs_text_encoder,
+                    return_unused_kwargs=True,
```
Comment (ylacombe) [inline src/transformers/models/musicgen/modeling_musicgen.py:1947]: Same here and for all the other changes below

**Review by ylacombe — APPROVED** [review]
Hey @staghado , thanks for iterating here ! And sorry for the delay!

There's still some changes that appeared because of added commas, but otherwise looks good to me!

cc @amyeroberts for a review! (btw, there's no need for additional tests right ?)

**Pull request closed without merging**